core/src/main/java/hudson/model/DescriptorVisibilityFilter.java

```diff
@@ -68,4 +93,29 @@
         }
         return r;
     }
+
+    public static <T extends Descriptor> List<T> applyType(Class<?> contextClass, Iterable<T> source) {
+        ExtensionList<DescriptorVisibilityFilter> filters = all();
+        List<T> r = new ArrayList<T>();
+
+        OUTER:
+        for (T d : source) {
+            if (LOGGER.isLoggable(Level.FINE)) {
+                LOGGER.fine("Determining visibility of " + d + " in contexts of type " + contextClass);
+            }
+            for (DescriptorVisibilityFilter f : filters) {
+                if (LOGGER.isLoggable(Level.FINER)) {
+                    LOGGER.finer("Querying " + f + " for visibility of " + d + " in type " + contextClass);
+                }
+                if (contextClass != null && !f.filterType(contextClass, d)) {
```

Comment:
Filter is an extension point, so maybe we want to be protected against Runtime exceptions here